peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+        bar: Annotated[ReadOnly[int], Gt(0)]
+
+.. code-block:: python
+
+    class Foo:
+        foo: ReadOnly[ClassVar[str]] = "foo"
+        bar: ReadOnly[Annotated[int, Gt(0)]]
+
+This is consistent with the interaction of ``ReadOnly`` and :class:`typing.TypedDict`
+defined in :pep:`705`.
+
+``ClassVar`` excludes read-only attributes from being assignable to within
+initialization methods.
+
+Rules of ``Final`` should take priority when combined with ``ReadOnly``. As such,
+type checkers may warn on the redundancy of combining the two type qualifiers.
+
+
+Backwards Compatibility
+=======================
+
+This PEP introduces new contexts where ``ReadOnly`` is valid. Programs inspecting
+those places will have to change to support it. This is expected to mainly affect type checkers.
+
+However, caution is advised while using the backported ``typing_extensions.ReadOnly``
+in older versions of Python. Mechanisms inspecting annotations may behave incorrectly
+when encountering ``ReadOnly``; in particular, the ``@dataclass`` decorator
+which `looks for <https://docs.python.org/3/library/dataclasses.html#class-variables>`_
+``ClassVar`` will incorrectly treat ``ReadOnly[ClassVar[...]]`` as an instance attribute.
+
+
+Security Implications
+=====================
+
+There are no known security consequences arising from this PEP.
+
+
+How to Teach This
+=================
+
+[How to teach users, new and experienced, how to apply the PEP to their work.]
+
+
+Open Issues
+===========
+
+Assignment in ``__new__``
+-------------------------
+
+Immutable classes like :class:`fractions.Fraction` often do not define ``__init__``;
+instead, they perform initialization in ``__new__`` or classmethods. The proposed
+feature won't be useful to them. 
+
+OTOH, allowing assignment within ``__new__`` (and/or classmethods) could open way
+to non-trivial bugs:
+
+.. code-block:: python
+
+    class Foo:
+        # fully initialized objects
+        object_cache: ReadOnly[ClassVar[dict[int, Self]]] = {}
+
+        foo: ReadOnly[int]
+
+        def __new__(cls, foo: int) -> Self:
+            if foo + 1 in cls.object_cache:
+                # this instance is already initialized
+                self = cls.object_cache[foo + 1]
+
+            else:
+                # this instance is not
+                self = super().__new__(cls)
+
+            # assignment to an object which has been initialized before,
+            # breaking the invariant a read-only attribute can be assigned to
+            # only during its initialization?
+            self.foo = foo
+
+            cls.object_cache[foo] = self
+            return self
+
+To my understanding, properly detecting this problem would require type checkers
+to keep track of the "level of initialization" of an object.
+
+This issue doesn't seem to impact ``__init__``, since it's rather uncommon to
+ever rebind ``self`` within it to any other object, and type checkers could
+flag the action as whole.
```

Comment:
This should probably also mention "calling `__init__` on an existing instance" as another action that would wrongly bypass readonly, if not flagged by type checkers.

Reply:
@Enegg I think this is still unresolved.

Reply:
Right. @carljm, since this issue has been removed, do you think there's still value in including what you said?
If yes, it'd probably be a note in the `Initialization` section, or perhaps we should view it as an Open Issue

Reply:
I think it would still be useful to note, in the Initialization section, that if type checkers allow explicit re-calling of the `__init__` method on an instance, this opens the door for a `ReadOnly` attribute to be reassigned on an existing instance.

Personally I would be OK with strengthening this to "type checkers should not allow re-calling `__init__` on an existing instance", but I would want to get feedback from @erictraut on that, since pyright currently does allow it.

Reply:
When you say "re-calling `__init__`", do you mean something like `super().__init__(<args>)`? This is a common pattern, so I don't think it should be disallowed in general. Are you suggesting that it be disallowed only if a `ReadOnly` field is present in one of the class' ancestors? I'm not in favor of mandating such a restriction. I'd want to see reasonably strong evidence that this is a common source of bugs before going down that road.

Reply:
@erictraut Sorry, no, this is what I mean:

```py
class A:
    def __init__(self, val: int) -> None:
        self.val: ReadOnly[int] = val

a = A(3)
a.__init__(4)
```

I think this is very rarely used, and it's already unsound with inheritance because we don't enforce LSP on `__init__`, so I think it shouldn't be allowed. But it's also probably rare enough that it doesn't matter too much that we require disallowing it.

Reply:
Ah, I see what you mean. 

IMO, I don't think this is a case that's important enough to cover in the typing spec one way or another. If someone wants to call `__init__` manually on an object, they're already treading on thin ice and had better know what they're doing.

From my perspective, it's not the job of a type checker to protect developers from all possible ways of misusing or abusing the language. This particular case feels more like a linter rule than something a type checker should check for.

Reply:
I don't feel strongly about specifying it. I don't even feel strongly about mentioning it in this PEP. Happy with whatever you prefer, @Enegg.